### ISSUE TITLE
NaN output all the time.

### ISSUE THREAD

**Issue by dionisole** [deeplearning4j/deeplearning4j#4303]
package org.deeplearning4j.examples.convolution;

import org.datavec.api.records.reader.RecordReader;
import org.datavec.api.records.reader.impl.csv.CSVRecordReader;
import org.datavec.api.split.FileSplit;
import org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator;
import org.deeplearning4j.eval.Evaluation;
import org.deeplearning4j.nn.api.OptimizationAlgorithm;
import org.deeplearning4j.nn.conf.LearningRatePolicy;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.conf.Updater;
import org.deeplearning4j.nn.conf.inputs.InputType;
import org.deeplearning4j.nn.conf.layers.ConvolutionLayer;
import org.deeplearning4j.nn.conf.layers.DenseLayer;
import org.deeplearning4j.nn.conf.layers.OutputLayer;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.deeplearning4j.optimize.listeners.ScoreIterationListener;
import org.deeplearning4j.ui.standalone.ClassPathResource;
import org.nd4j.linalg.activations.Activation;
import org.nd4j.linalg.dataset.DataSet;
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator;
import org.nd4j.linalg.dataset.api.preprocessor.DataNormalization;
import org.nd4j.linalg.dataset.api.preprocessor.NormalizerStandardize;
import org.nd4j.linalg.lossfunctions.LossFunctions;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;

public class p5 {

    private static Logger log = LoggerFactory.getLogger(p5.class);

    public static void main(String[] args) throws  Exception {
        //First: get the dataset using the record reader. CSVRecordReader handles loading/parsing
        int numLinesToSkip = 0;
        String delimiter = ",";
        RecordReader recordReader = new CSVRecordReader(numLinesToSkip,delimiter);
        recordReader.initialize(new FileSplit(new ClassPathResource("protein_csv_train_all.txt").getFile()));
        RecordReader recordReader1 = new CSVRecordReader(numLinesToSkip,delimiter);
        recordReader1.initialize(new FileSplit(new ClassPathResource("protein_csv_test_all.txt").getFile()));

        //Second: the RecordReaderDataSetIterator handles conversion to DataSet objects, ready for use in neural network
        int labelIndex = 20;     //5 values in each row of the iris.txt CSV: 4 input features followed by an integer label (class) index. Labels are the 5th value (index 4) in each row
        int numClasses = 3;     //3 classes (types of iris flowers) in the iris data set. Classes have integer values 0, 1 or 2
        int trainbatchSize = 6000;    //31286 examples total. We are loading all of them into one DataSet (not recommended for large data sets)
        int testBatchSize=300; // 15695 total
        int numOfTestBatches=25;
        int numOfTrainBatches=13;
        int epochs=20;
        DataSetIterator iterator = new RecordReaderDataSetIterator(recordReader,trainbatchSize,labelIndex,numClasses,numOfTrainBatches);
        DataSetIterator iterator1 = new RecordReaderDataSetIterator(recordReader1,testBatchSize,labelIndex,numClasses,numOfTestBatches);

        //We need to normalize our data. We'll use NormalizeStandardize (which gives us mean 0, unit variance):
        DataNormalization scaler = new NormalizerStandardize();
        scaler.fit(iterator);
        iterator.setPreProcessor(scaler);

        DataNormalization scaler1 = new NormalizerStandardize();
        scaler1.fit(iterator1);
        iterator1.setPreProcessor(scaler1);

        final int numInputs = 20;
        int outputNum = 3;
        int iterations = 1;
        int nChannels = 1;

        log.info("Build model....");
        // learning rate schedule in the form of <Iteration #, Learning Rate>
        Map<Integer, Double> lrSchedule = new HashMap<>();
        lrSchedule.put(0, 0.05);
        lrSchedule.put(26,0.03);
        lrSchedule.put(52,0.008);
        lrSchedule.put(78,0.004);

        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .iterations(iterations) // Training iterations as above
                .regularization(true).l2(0.005)
                .learningRateDecayPolicy(LearningRatePolicy.Schedule)
                .learningRateSchedule(lrSchedule)
                .weightInit(WeightInit.RELU)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(Updater.NESTEROVS).momentum(0.8)
                .list()
                .layer(0, new ConvolutionLayer.Builder()
                        .kernelSize(2,2)
                        .stride(1,1)
                        //nIn and nOut specify depth. nIn here is the nChannels and nOut is the number of filters to be applied
                        .nIn(nChannels) //for ConvolutionLayer
                        //.stride(1)
                        .padding(2,2)
                        //.padding(2)
                        .nOut(10)
                        .activation(Activation.RELU)
                        .build())
                .layer(1, new ConvolutionLayer.Builder()
                        .kernelSize(2,2)
                        .stride(1,1)
                        //Note that nIn need not be specified in later layers
                        //.stride(1, 1)
                        .nIn(10)
                        .padding(2,2)
                        //.stride(1)
                        //.padding(2)
                        .nOut(20)
                        .activation(Activation.RELU)
                        .build())
                .layer(2, new ConvolutionLayer.Builder()
                        .kernelSize(2,2)
                        .stride(1,1)
                        //Note that nIn need not be specified in later layers
                        //.stride(1, 1)
                        .nIn(20)
                        .padding(2,2)
                        //.stride(1)
                        //.padding(2)
                        .nOut(10)
                        .activation(Activation.RELU)
                        .build())
                .layer(3, new ConvolutionLayer.Builder()
                        .kernelSize(2,2)
                        .stride(1,1)
                        //Note that nIn need not be specified in later layers
                        //.stride(1, 1)
                        .nIn(10)
                        .padding(2,2)
                        //.stride(1)
                        //.padding(2)
                        .nOut(10)
                        .activation(Activation.RELU)
                        .build())
                .layer(4, new DenseLayer.Builder().nOut(3).build())
                .layer(5, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .nIn(3)
                        .nOut(outputNum)
                        .activation(Activation.SOFTMAX)
                        .build())
                .setInputType(InputType.convolutionalFlat(1,20,1)) //See note below
                .backprop(true).pretrain(false).build();

        //run the model
        MultiLayerNetwork model = new MultiLayerNetwork(conf);
        model.init();

        //TRAINING
        model.setListeners(new ScoreIterationListener(1)); //comment for UI VISUALIZATION
        System.out.println("START TRAINING!");
        DataSet trainNextSet=null;
        for (int j=0;j<epochs;j++) {
            //System.out.println("EPOCH "+j);
            for (int i = 0; i < numOfTrainBatches; i++) {
                if (iterator.hasNext()) {
                    trainNextSet = iterator.next();
                    model.fit(trainNextSet);
                }
            }
            iterator.reset();
        }

        //TESTING
        System.out.println("START TESTING!");
        Evaluation a = model.evaluate(iterator1);
        System.out.println(a.stats());
        System.out.println("END OF TESTING!");

    }

}

**Comment by AlexDBlack** [comment]
Closing both of these issues. Will discuss in tuning channel.
